src/csharp/Grpc.Core/Internal/ReusableSliceBuffer.cs

```diff
@@ -0,0 +1,148 @@
+
+#if GRPC_CSHARP_SUPPORT_SYSTEM_MEMORY
+
+using Grpc.Core.Utils;
+using System;
+using System.Threading;
+
+using System.Buffers;
+
+namespace Grpc.Core.Internal
+{
+    internal class ReusableSliceBuffer
+    {
+        public const int MaxCachedSegments = 1024;  // ~4MB payload for 4K slices
+
+        readonly SliceSegment[] cachedSegments = new SliceSegment[MaxCachedSegments];
+        int populatedSegmentCount = 0;
+
+        public ReadOnlySequence<byte> PopulateFrom(IBufferReader bufferReader)
+        {
+            long offset = 0;
+            int index = 0;
+            SliceSegment prevSegment = null;
+            while (bufferReader.TryGetNextSlice(out Slice slice))
+            {
+                // Initialize cached segment if still null or just allocate a new segment if we already reached MaxCachedSegments
+                var current = index < cachedSegments.Length ? cachedSegments[index] : new SliceSegment();
+                if (current == null)
+                {
+                    current = cachedSegments[index] = new SliceSegment();
+                }
+
+                current.Reset(slice, offset);
+                prevSegment?.SetNext(current);
+
+                index ++;
+                offset += slice.Length;
+                prevSegment = current;
+            }
+            populatedSegmentCount = index;
+
+            // Not necessary for ending the ReadOnlySequence, but for making sure we
+            // don't keep more than MaxCachedSegments alive.
+            prevSegment?.SetNext(null);
+
+            if (index == 0)
+            {
+                return ReadOnlySequence<byte>.Empty;
+            }
+
+            var firstSegment = cachedSegments[0];
+            var lastSegment = prevSegment;
+            return new ReadOnlySequence<byte>(firstSegment, 0, lastSegment, lastSegment.Memory.Length);
+        }
+
+        public void Invalidate()
+        {
+            if (populatedSegmentCount == 0)
+            {
+                return;
+            }
+            var segment = cachedSegments[0];
+            while (segment != null)
+            {
+                segment.Reset(new Slice(IntPtr.Zero, 0), 0);
+                segment.SetNext(null);
+                segment = (SliceSegment) segment.Next;
+            }
+            populatedSegmentCount = 0;
+        }
+
+        // Represents a segment in ReadOnlySequence
+        // Segment is backed by Slice and the instances are reusable.
+        private class SliceSegment : ReadOnlySequenceSegment<byte>
+        {
+            readonly SliceMemoryManager pointerMemoryManager = new SliceMemoryManager();
+
+            public void Reset(Slice slice, long runningIndex)
+            {
+                pointerMemoryManager.Reset(slice);
+                Memory = pointerMemoryManager.Memory;  // maybe not always necessary
+                RunningIndex = runningIndex;
+            }
+
+            public void SetNext(ReadOnlySequenceSegment<byte> next)
```

Comment:
nit: IMO this setter reduces readability. Can we remove it and just access `Next` directly, since IIC it's already public?